src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java

```diff
@@ -79,17 +112,141 @@ enum Signal {
    */
   final Map<RepoRecordedInput, String> recordedInputValues = new TreeMap<>();
 
+  RepoFetchingSkyKeyComputeState(String repositoryName,
+      Function<RepoFetchingSkyKeyComputeState, Runnable> threadFunc) {
+    this.repositoryName = repositoryName;
+    workerThread =
+        Thread.ofVirtual()
+            .name("starlark-repository-" + repositoryName)
+            .start(threadFunc.apply(this));
+  }
+
   SkyFunction.Environment signalForFreshEnv() throws InterruptedException {
-    signalQueue.put(Signal.RESTART);
-    return delegateEnvQueue.take();
+    Response response = sendRequest(new Request.NewDependencies());
+    if (!(response instanceof Response.Restarting)) {
```

Comment:
Since `close` can be called at anytime, the request queue might be drained by the message loop and we can possibly get back `FailureAcknowledged`. I think this applies to every call sites of `sendRequest[Uninterruptibly]`.

Reply:
Correct. I thought that was OK because then the fate of the worker thread doesn't matter anymore. If you insist, I can add some code to handle the "close is discarding everything" case, though.

Reply:
Actually no, that can't happen -- this is the exact condition the sequence number protects against: `sendRequest()` matches up the requests with the responses, so if a given `sendRequest()` call is given a `NewDependencies` as an argument it can only ever get back `Restarting`.

Reply:
Actually-actually, you're right; let me just fix this.